internal/engine/selectors/selectors.go

```diff
@@ -0,0 +1,370 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate go run go.uber.org/mock/mockgen -package mock_$GOPACKAGE -destination=./mock/$GOFILE -source=./$GOFILE
+
+// Package selectors provides utilities for selecting entities based on profiles using CEL
+package selectors
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/interpreter"
+
+	internalpb "github.com/stacklok/minder/internal/proto"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+var (
+	// ErrResultUnknown is returned when the result of a selector expression is unknown
+	// this tells the caller to try again with more information
+	ErrResultUnknown = errors.New("result is unknown")
+)
+
+// celEnvFactory is an interface for creating CEL environments
+// for an entity. Each entity must implement this interface to be
+// usable in selectors
+type celEnvFactory func() (*cel.Env, error)
+
+// genericEnvFactory is a factory for creating a CEL environment
+// for the generic SelectorEntity type
+func genericEnvFactory() (*cel.Env, error) {
+	return newEnvForEntity(
+		"entity",
+		&internalpb.SelectorEntity{},
+		"internal.SelectorEntity")
+}
+
+// repoEnvFactory is a factory for creating a CEL environment
+// for the SelectorRepository type representing a repository
+func repoEnvFactory() (*cel.Env, error) {
+	return newEnvForEntity(
+		"repository",
+		&internalpb.SelectorRepository{},
+		"internal.SelectorRepository")
+}
+
+// artifactEnvFactory is a factory for creating a CEL environment
+// for the SelectorArtifact type representing an artifact
+func artifactEnvFactory() (*cel.Env, error) {
+	return newEnvForEntity(
+		"artifact",
+		&internalpb.SelectorArtifact{},
+		"internal.SelectorArtifact")
+}
+
```

Comment:
Do we need an env factory for pull requests?

Reply:
if you can come up with resonable selector message for the pull request, sure. I wasn't sure what to put there honestly. Maybe we could just have properties and nothing else?

Reply:
Let's add a TODO comment for this. I think it's just fine if it's just properties.

Reply:
done

Reply:
Where? :confused:

Reply:
sorry, only to the proto message definition. I added another one to the env factory, too